### PR TITLE
payload and compensation saving in saga log

### PR DESCRIPTION
**Pull request by SzymonWelter**

### CODE REVIEW

File: Orchestrator/Prognetics.Saga.Orchestrator/Prognetics.Saga.Orchestrator.csproj [hunk 1]
```diff
@@ -28,6 +28,10 @@
     </None>
   </ItemGroup>
 
+  <ItemGroup>
+    <PackageReference Include="Microsoft.Extensions.Logging.Abstractions" Version="7.0.0" />
```
Comment (ChrisGawlinski) [inline Orchestrator/Prognetics.Saga.Orchestrator/Prognetics.Saga.Orchestrator.csproj:32]: If we want to add logger, please do so as part of separate PR.

File: Orchestrator/Prognetics.Saga.Orchestrator/SagaEngine.cs [hunk 2]
```diff
@@ -1,33 +1,125 @@
...
+        if (transactionStep.IsFirst)
+        {
+            transactionId = await InitializeTransaction(transactionStep);
+        }
+        else if (transactionId is null)
+        {
+            _logger.LogError("Transaction id is missing");
+            return null;
+        }
+        else if (await TryProcessExistingTransaction(transactionId, transactionStep) == false)
```
Comment (ChrisGawlinski) [inline Orchestrator/Prognetics.Saga.Orchestrator/SagaEngine.cs:54]: Order of those if's is hardly readable. It's hard to tell why next steps occur in this specific way. Can we maybe split this into more methods or use some pattern?

File: Orchestrator/Prognetics.Saga.Orchestrator/SagaEngine.cs [hunk 3]
```diff
@@ -1,33 +1,125 @@
...
+        var transactionId = input.Message.TransactionId;
+
+        if (transactionStep.IsFirst)
+        {
+            transactionId = await InitializeTransaction(transactionStep);
+        }
+        else if (transactionId is null)
+        {
+            _logger.LogError("Transaction id is missing");
+            return null;
```
Comment (ChrisGawlinski) [inline Orchestrator/Prognetics.Saga.Orchestrator/SagaEngine.cs]: What does it mean if we return null? Didn't we want to avoid using nulls?

File: Orchestrator/Prognetics.Saga.Orchestrator/SagaEngine.cs [hunk 4]
```diff
@@ -1,33 +1,125 @@
...
 public class SagaEngine : ISagaEngine
 {
+    private readonly ISagaLog _sagaLog;
+    private readonly ICompensationStore _compensationStore;
     private readonly ITransactionLedgerAccessor _transactionLedgerAccessor;
+    private readonly IdGenerator _idGenerator;
+    private readonly ILogger<ISagaEngine> _logger;
 
-    public SagaEngine(ITransactionLedgerAccessor transactionLedgerAccessor)
+    public SagaEngine(
```
Comment (ChrisGawlinski) [inline Orchestrator/Prognetics.Saga.Orchestrator/SagaEngine.cs:18]: Is this tested?

File: Orchestrator/Prognetics.Saga.Orchestrator/SagaEngine.cs (hunk 3 above)
Comment (SzymonWelter) [reply Orchestrator/Prognetics.Saga.Orchestrator/SagaEngine.cs]: I see a few ways of dealing with invalid messages:
- returning null if processing have failed, the method name has a postfix 'OrDefault'
- returning bool and the engine output as an out parameter
- throwing an exception
- IsSuccess prop in the engine output
I've chosen the first one because I like dealing with nulls with the nullable feature enabled in .NET. However i'm open to the other solutions as well.

File: Orchestrator/Prognetics.Saga.Orchestrator/Prognetics.Saga.Orchestrator.csproj (hunk 1 above)
Comment (SzymonWelter) [reply Orchestrator/Prognetics.Saga.Orchestrator/Prognetics.Saga.Orchestrator.csproj:32]: Its not a specific logger. Its a microsoft abstraction of logging that every modern dotnet logging framework implements. We can create a logging abstraction ourselves but I don't see value in that.

File: Orchestrator/Prognetics.Saga.Orchestrator/SagaEngine.cs (hunk 3 above)
Comment (ChrisGawlinski) [reply Orchestrator/Prognetics.Saga.Orchestrator/SagaEngine.cs]: As discussed let's go with the engine output :)

**Review by ChrisGawlinski — APPROVED** [review]
